From 4feab29fb11068383b50577b576db41ef5020701 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Sat, 3 Mar 2018 10:10:01 +0100
Subject: [PATCH] [mpp_dec]: sleep when there is nothing to parse

---
 mpp/codec/mpp_dec.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/mpp/codec/mpp_dec.cpp b/mpp/codec/mpp_dec.cpp
index 5bb59534..a0be80ea 100644
--- a/mpp/codec/mpp_dec.cpp
+++ b/mpp/codec/mpp_dec.cpp
@@ -725,8 +725,10 @@ void *mpp_dec_parser_thread(void *data)
             continue;
         }
 
-        if (try_proc_dec_task(mpp, &task))
+        if (try_proc_dec_task(mpp, &task)) {
+            msleep(1);
             continue;
+        }
 
     }
 
